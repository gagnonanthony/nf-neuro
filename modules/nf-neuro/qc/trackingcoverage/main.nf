process QC_TRACKINGCOVERAGE {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:latest' }"

    input:
        tuple val(meta), path(tractogram), path(mask)

    output:
        tuple val(meta), path("*__tractogram_mask.nii.gz")   , emit: tractogram_mask
        tuple val(meta), path("*__TDI.nii.gz")               , emit: TDI
        tuple val(meta), path("*__stats.json")               , emit: stats
        tuple val(meta), path("*__sc.txt")                   , emit: sc
        path "*.png"                                         , emit: png
        path "versions.yml"                                  , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def sh_basis = task.ext.sh_basis ? "--sh_basis ${task.ext.sh_basis}" : ''
    def sphere = task.ext.sphere ? "--sphere ${task.ext.sphere}" : ''
    def sh_order = task.ext.sh_order ? "--sh_order ${task.ext.sh_order}" : ''
    def normalize_per_voxel = task.ext.normalize_per_voxel ? "--normalize_per_voxel" : ''
    def smooth_todi = task.ext.smooth_todi ? "--smooth_todi" : ''
    def asymmetric = task.ext.asymmetric ? "--asymmetric" : ''
    def n_steps = task.ext.n_steps ? "--n_steps ${task.ext.n_steps}" : ''

    """
    scil_tractogram_count_streamlines.py $tractogram --print_count_alone > ${prefix}__sc.txt

    # Computing TODI.
    scil_tractogram_compute_TODI.py $tractogram \
        --out_mask ${prefix}__tractogram_mask.nii.gz \
        --out_tdi ${prefix}__TDI.nii.gz \
        $sh_basis $sphere $sh_order $normalize_per_voxel \
        $smooth_todi $asymmetric $n_steps

    # Computing DICE score.
    scil_volume_pairwise_comparison.py $mask ${prefix}__tractogram_mask.nii.gz \
        ${prefix}__stats.json

    # Fetch middle axial slice.
    size=\$(mrinfo ${prefix}__TDI.nii.gz -size)
    mid_slice=\$(echo \$size | awk '{print int((\$2 + 1) / 2)}')

    # Visual QC file.
    scil_viz_volume_screenshot.py ${prefix}__TDI.nii.gz ${prefix}__coverage_overlay.png \
        --volume_cmap pink \
        --overlays $mask \
        --overlays_opacity 0 \
        --overlays_as_contours \
        --display_slice_number \
        --display_lr \
        --overlays_colors 0 255 0 \
        --slices \$mid_slice \
        --axis axial

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list --disable-pip-version-check --no-python-version-warning | grep scilpy | tr -s ' ' | cut -d ' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    touch ${prefix}__tractogram_mask.nii.gz
    touch ${prefix}__TDI.nii.gz
    touch ${prefix}__stats.json
    touch ${prefix}__sc.txt
    touch ${prefix}__coverage_overlay.png

    scil_tractogram_count_streamlines.py -h
    scil_tractogram_compute_TODI.py -h
    scil_volume_pairwise_comparison.py -h
    scil_viz_volume_screenshot.py -h

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list --disable-pip-version-check --no-python-version-warning | grep scilpy | tr -s ' ' | cut -d ' ' -f2)
    END_VERSIONS
    """
}
