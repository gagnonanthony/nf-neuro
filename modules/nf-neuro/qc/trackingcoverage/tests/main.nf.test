nextflow_process {

    name "Test Process QC_TRACKINGCOVERAGE"
    script "../main.nf"
    process "QC_TRACKINGCOVERAGE"

    tag "modules"
    tag "modules_nfcore"
    tag "qc"
    tag "qc/trackingcoverage"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "tracking.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("qc - trackingcoverage") {

        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/pft.trk"),
                        file("\${test_data_directory}/seeding_mask.nii.gz")
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
