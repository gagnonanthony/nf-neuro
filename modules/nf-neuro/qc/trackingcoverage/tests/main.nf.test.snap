{
    "qc - trackingcoverage": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__tractogram_mask.nii.gz:md5,4ce57fee196997a972d5c33202eeb74f"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__TDI.nii.gz:md5,f4d9df11957de7b4b77314e07e08afa2"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__stats.json:md5,9078688e95816d3479936e74b17a2d21"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__sc.txt:md5,dc815a1f602788e79d9e5dbcc15b05fc"
                    ]
                ],
                "4": [
                    "test__coverage_overlay_slice_42.png:md5,1c031c38485271e5ba755947f9644c80"
                ],
                "5": [
                    "versions.yml:md5,bac162443aba01f87005a66902dfee95"
                ],
                "TDI": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__TDI.nii.gz:md5,f4d9df11957de7b4b77314e07e08afa2"
                    ]
                ],
                "png": [
                    "test__coverage_overlay_slice_42.png:md5,1c031c38485271e5ba755947f9644c80"
                ],
                "sc": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__sc.txt:md5,dc815a1f602788e79d9e5dbcc15b05fc"
                    ]
                ],
                "stats": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__stats.json:md5,9078688e95816d3479936e74b17a2d21"
                    ]
                ],
                "tractogram_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__tractogram_mask.nii.gz:md5,4ce57fee196997a972d5c33202eeb74f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bac162443aba01f87005a66902dfee95"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-07T15:59:12.176208"
    }
}